From 3ae89bc08319c0e9a4e2d72cbf2d332e87486420 Mon Sep 17 00:00:00 2001
From: Luciana Fujii Pontello <luciana@fujii.eti.br>
Date: Mon, 12 Dec 2011 20:21:54 -0200
Subject: [PATCH 129/154] pvrvideosink: Send application message with crop
 size

Sends the original resolution of the video (not the padded one) as an
application message. This allows applications to use the right
resolution for any purpose, such as calculating aspect ratio themselves
as totem does.
---
 sys/pvr2d/gstpvrvideosink.c |   21 +++++++++++++++++----
 1 files changed, 17 insertions(+), 4 deletions(-)

diff --git a/sys/pvr2d/gstpvrvideosink.c b/sys/pvr2d/gstpvrvideosink.c
index c65af76..abb911f 100644
--- a/sys/pvr2d/gstpvrvideosink.c
+++ b/sys/pvr2d/gstpvrvideosink.c
@@ -1096,6 +1096,8 @@ gst_pvrvideosink_event (GstBaseSink * bsink, GstEvent * event)
 {
   gboolean res;
   GstPVRVideoSink *pvrvideosink = GST_PVRVIDEOSINK (bsink);
+  GstStructure *structure;
+  GstMessage *message;
 
   switch (GST_EVENT_TYPE (event)) {
     case GST_EVENT_CROP:
@@ -1106,15 +1108,26 @@ gst_pvrvideosink_event (GstBaseSink * bsink, GstEvent * event)
       gst_event_parse_crop (event, &top, &left, &width, &height);
       c->top = top;
       c->left = left;
-      if (width == -1)
+      if (width == -1) {
         c->right = GST_VIDEO_SINK_WIDTH (pvrvideosink);
-      else
+        width = GST_VIDEO_SINK_WIDTH (pvrvideosink);
+      } else {
         c->right = left + width;
+      }
 
-      if (height == -1)
+      if (height == -1) {
         c->bottom = GST_VIDEO_SINK_HEIGHT (pvrvideosink);
-      else
+        height = GST_VIDEO_SINK_HEIGHT (pvrvideosink);
+      } else {
         c->bottom = top + height;
+      }
+
+      structure = gst_structure_new ("video-size-crop", "width", G_TYPE_INT,
+          width, "height", G_TYPE_INT, height, NULL);
+      message = gst_message_new_application (GST_OBJECT (pvrvideosink),
+          structure);
+      gst_bus_post (gst_element_get_bus (GST_ELEMENT (pvrvideosink)), message);
+
 
       if (!gst_pvrvideosink_configure_overlay (pvrvideosink, width, height,
               pvrvideosink->video_par_n, pvrvideosink->video_par_d,
-- 
1.7.5.4

